File: polyglot/go/mesh/mesh.go
// Package mesh implements UDP mesh networking for the unit protocol.
package mesh

import (
	"crypto/rand"
	"fmt"
	"log"
	"net"
	"sync"

	"github.com/DavidCanHelp/unit/polyglot/go/sexp"
)

// Peer represents a known mesh peer.
type Peer struct {
	ID      string
	Addr    *net.UDPAddr
	Fitness int64
	Energy  int64
}

// Mesh manages UDP mesh communication.
type Mesh struct {
	ID    string
	Port  int
	conn  *net.UDPConn
	mu    sync.Mutex
	peers map[string]*Peer // keyed by ID
	OnMsg func(sexp.SExp, *net.UDPAddr)
}

// NewMesh creates a mesh node. If id is empty, generates a random one.
func NewMesh(id string, port int) (*Mesh, error) {
	if id == "" {
		b := make([]byte, 4)
		rand.Read(b)
		id = fmt.Sprintf("%02x%02x%02x%02x", b[0], b[1], b[2], b[3])
	}

	addr, err := net.ResolveUDPAddr("udp", fmt.Sprintf("0.0.0.0:%d", port))
	if err != nil {
		return nil, err
	}
	conn, err := net.ListenUDP("udp", addr)
	if err != nil {
		return nil, err
	}

	return &Mesh{
		ID:    id,
		Port:  port,
		conn:  conn,
		peers: make(map[string]*Peer),
	}, nil
}

// Listen starts receiving UDP messages in a goroutine.
func (m *Mesh) Listen() {
	go func() {
		buf := make([]byte, 65536)
		for {
			n, addr, err := m.conn.ReadFromUDP(buf)
			if err != nil {
				continue
			}
			msg := string(buf[:n])
			parsed, err := sexp.Parse(msg)
			if err != nil {
				continue
			}
			m.handleMessage(parsed, addr)
			if m.OnMsg != nil {
				m.OnMsg(parsed, addr)
			}
		}
	}()
}

func (m *Mesh) handleMessage(s sexp.SExp, addr *net.UDPAddr) {
	switch s.MsgType() {
	case "peer-announce", "peer-status":
		idNode := s.GetKeyword("id")
		if idNode == nil {
			return
		}
		peerID, ok := idNode.AsStr()
		if !ok {
			return
		}
		if peerID == m.ID {
			return // ignore self
		}
		m.mu.Lock()
		p, exists := m.peers[peerID]
		if !exists {
			p = &Peer{ID: peerID, Addr: addr}
			m.peers[peerID] = p
			log.Printf("[mesh] discovered peer %s @ %s", peerID, addr)
		}
		p.Addr = addr
		if f := s.GetKeyword("fitness"); f != nil {
			p.Fitness, _ = f.AsInt()
		}
		if e := s.GetKeyword("energy"); e != nil {
			p.Energy, _ = e.AsInt()
		}
		m.mu.Unlock()
	}
}

// Send sends a raw S-expression string to a specific address.
func (m *Mesh) Send(addr *net.UDPAddr, msg string) {
	m.conn.WriteToUDP([]byte(msg), addr)
}

// SendToAll broadcasts an S-expression string to all known peers.
func (m *Mesh) SendToAll(msg string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	for _, p := range m.peers {
		m.conn.WriteToUDP([]byte(msg), p.Addr)
	}
}

// Announce sends a peer-announce to a specific address.
func (m *Mesh) Announce(peerAddr string) error {
	addr, err := net.ResolveUDPAddr("udp", peerAddr)
	if err != nil {
		return err
	}
	msg := fmt.Sprintf(`(peer-announce :id "%s" :port %d)`, m.ID, m.Port)
	m.Send(addr, msg)
	return nil
}

// GossipTick broadcasts peer-status to all known peers.
func (m *Mesh) GossipTick(fitness, energy int64) {
	m.mu.Lock()
	peerCount := len(m.peers)
	m.mu.Unlock()
	msg := fmt.Sprintf(
		`(peer-status :id "%s" :peers %d :fitness %d :energy %d)`,
		m.ID, peerCount, fitness, energy,
	)
	m.SendToAll(msg)
}

// PeerCount returns the number of known peers.
func (m *Mesh) PeerCount() int {
	m.mu.Lock()
	defer m.mu.Unlock()
	return len(m.peers)
}

// Close shuts down the mesh.
func (m *Mesh) Close() {
	m.conn.Close()
}

File: polyglot/python/mesh.py
"""UDP mesh networking for the unit protocol."""

import os
import socket
import threading
import logging

import sexp

log = logging.getLogger(__name__)


class Peer:
    def __init__(self, id, addr):
        self.id = id
        self.addr = addr
        self.fitness = 0
        self.energy = 0


class UDPMesh:
    def __init__(self, node_id=None, port=4202):
        if node_id is None:
            node_id = os.urandom(4).hex()
        self.id = node_id
        self.port = port
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.sock.bind(('0.0.0.0', port))
        self.sock.settimeout(1.0)
        self.peers = {}  # id -> Peer
        self.on_message = None  # callback(parsed, addr)
        self._lock = threading.Lock()

    def listen(self):
        t = threading.Thread(target=self._recv_loop, daemon=True)
        t.start()

    def _recv_loop(self):
        while True:
            try:
                data, addr = self.sock.recvfrom(65536)
                msg = data.decode('utf-8', errors='replace')
                parsed = sexp.parse(msg)
                self._handle(parsed, addr)
                if self.on_message:
                    self.on_message(parsed, addr)
            except socket.timeout:
                continue
            except Exception:
                continue

    def _handle(self, parsed, addr):
        mt = sexp.msg_type(parsed)
        if mt in ('peer-announce', 'peer-status'):
            peer_id = sexp.get_keyword(parsed, 'id')
            if not peer_id or peer_id == self.id:
                return
            with self._lock:
                if peer_id not in self.peers:
                    self.peers[peer_id] = Peer(peer_id, addr)
                    log.info(f"discovered peer {peer_id} @ {addr}")
                p = self.peers[peer_id]
                p.addr = addr
                f = sexp.get_keyword(parsed, 'fitness')
                if isinstance(f, int):
                    p.fitness = f
                e = sexp.get_keyword(parsed, 'energy')
                if isinstance(e, int):
                    p.energy = e

    def send(self, addr, msg):
        self.sock.sendto(msg.encode(), addr)

    def send_to_all(self, msg):
        with self._lock:
            for p in self.peers.values():
                self.sock.sendto(msg.encode(), p.addr)

    def announce(self, peer_addr):
        addr = _resolve(peer_addr)
        if addr:
            msg = f'(peer-announce :id "{self.id}" :port {self.port})'
            self.send(addr, msg)

    def gossip_tick(self, fitness, energy):
        with self._lock:
            n = len(self.peers)
        msg = f'(peer-status :id "{self.id}" :peers {n} :fitness {fitness} :energy {energy})'
        self.send_to_all(msg)

    def peer_count(self):
        with self._lock:
            return len(self.peers)

    def close(self):
        self.sock.close()


def _resolve(addr_str):
    try:
        host, port = addr_str.rsplit(':', 1)
        return (host, int(port))
    except Exception:
        return None
